mm2src/coins/eth/erc20.rs

```diff
@@ -0,0 +1,102 @@
+use crate::eth::web3_transport::Web3Transport;
+use crate::eth::{EthCoin, ERC20_CONTRACT};
+use crate::{CoinsContext, MmCoinEnum};
+use ethabi::Token;
+use ethereum_types::Address;
+use futures_util::TryFutureExt;
+use mm2_core::mm_ctx::MmArc;
+use mm2_err_handle::mm_error::MmResult;
+use web3::types::{BlockId, BlockNumber, CallRequest};
+use web3::{Transport, Web3};
+
+async fn call_erc20_function<T: Transport>(
+    web3: &Web3<T>,
+    token_addr: Address,
+    function_name: &str,
+) -> Result<Vec<Token>, String> {
+    let function = try_s!(ERC20_CONTRACT.function(function_name));
+    let data = try_s!(function.encode_input(&[]));
+    let request = CallRequest {
+        from: Some(Address::default()),
+        to: Some(token_addr),
+        gas: None,
+        gas_price: None,
+        value: Some(0.into()),
+        data: Some(data.into()),
+        ..CallRequest::default()
+    };
+
+    let res = web3
+        .eth()
+        .call(request, Some(BlockId::Number(BlockNumber::Latest)))
+        .map_err(|e| ERRL!("{}", e))
+        .await?;
+    function.decode_output(&res.0).map_err(|e| ERRL!("{}", e))
+}
+
+pub(crate) async fn get_token_decimals(web3: &Web3<Web3Transport>, token_addr: Address) -> Result<u8, String> {
+    let tokens = call_erc20_function(web3, token_addr, "decimals").await?;
+    match tokens[0] {
+        Token::Uint(dec) => Ok(dec.as_u64() as u8),
+        _ => ERR!("Invalid decimals type {:?}", tokens),
+    }
+}
+
+async fn get_token_symbol(coin: &EthCoin, token_addr: Address) -> Result<String, String> {
+    let web3 = try_s!(coin.web3().await);
+    let tokens = call_erc20_function(&web3, token_addr, "symbol").await?;
+    match &tokens[0] {
+        Token::String(symbol) => Ok(symbol.clone()),
+        _ => ERR!("Invalid symbol type {:?}", tokens),
+    }
+}
```

Comment:
I wouldn't try to access index 0 directly as it would panic when `tokens` is empty.

Reply:
unfortunately, we seem to have these all over the eth code (e.g. `\[\d\]` in `eth_taker_swap_v2.rs`), and we consider this safe for some reason :/.
i verified how such vecs come and their sizes in a previous PR, but i would still be on the safe side and not have them addressed like that. also reducing tragic effects from dev errors.

Reply:
> I wouldn't try to access index 0 directly as it would panic when tokens is empty.

Missed this when copying the code. Will fix only the instances in this PR and fix all others in another PR.

Reply:
Fixed here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2141/commits/f38fc3fd457a56a2613ed26df686d922d6954a04